components/support/nimbus-fml/src/client/inspector.rs

```diff
@@ -406,3 +408,481 @@ mod unit_tests {
         Ok(())
     }
 }
+
+#[cfg(test)]
+mod correction_candidates {
+    use crate::{
+        client::test_helper::client,
+        editing::{CorrectionCandidate, CursorSpan},
+    };
+
+    use super::*;
+
+    // Makes a correction; this is a simulation of what the editor will do.
+    fn perform_correction(
+        lines: &[&str],
+        position: &CursorSpan,
+        correction: &CorrectionCandidate,
+    ) -> String {
+        let position = correction.insertion_span.as_ref().unwrap_or(position);
+        position.insert_str(lines, &correction.insert)
+    }
+
+    /// Takes an editor input and an inspector.
+    /// The editor input (lines) should have exactly one thing wrong with it.
+    ///
+    /// The correction candidates are tried one by one, and then the lines are
+    /// inspected again.
+    ///
+    /// The function fails if:
+    /// a) there are no errors in the initial text
+    /// b) there are no completions in the first error.
+    /// c) after applying each correction, then there is still an error.
+    ///
+    /// For obvious reasons, this does not handle arbitrary text. Some text will have too
+    /// many errors, some will not have any corrections, and some errors will not be corrected
+    /// by every correction (e.g. the key in a feature or object).
+    fn try_correcting_single_error(inspector: &FmlFeatureInspector, lines: &[&str]) {
+        let input = lines.join("\n");
+        let err = inspector.get_first_error(input.clone());
+        assert_ne!(None, err, "No error found in input: {input}");
+        let err = err.unwrap();
+        assert_ne!(
+            0,
+            err.corrections.len(),
+            "No corrections for {input}: {err:?}"
+        );
+
+        for correction in &err.corrections {
+            let input = perform_correction(lines, &err.error_span, correction);
+            let err = inspector.get_first_error(input.clone());
+            assert_eq!(None, err, "Error found in {input}");
+        }
+    }
```

Comment:
This method is the heart of these tests.

These tests test all the layers at the same time, so they are less brittle in the face of changing implementation details.